modules/core/src/main/java/org/apache/ignite/internal/processors/datastructures/DataStructuresProcessor.java

```diff
@@ -910,19 +914,33 @@ private CacheConfiguration metaCacheConfiguration(CollectionConfiguration cfg, S
     }
 
     /**
+     * Get compatible with collection configuration data structure cache.
+     *
      * @param cfg Collection configuration.
-     * @return Cache name.
      * @param grpName Group name.
+     * @param shared Shared mode (single cache for multiple data structure instances).
+     * @param dsName Data structure name.
+     * @return Data strcuture cache.
      * @throws IgniteCheckedException If failed.
      */
-    @Nullable private IgniteInternalCache compatibleCache(CollectionConfiguration cfg, String grpName)
-        throws IgniteCheckedException
-    {
+    @Nullable private IgniteInternalCache compatibleCache(CollectionConfiguration cfg, String grpName,
+        boolean shared, String dsName)
+        throws IgniteCheckedException {
+
+        // At first step we need to search shared compatible with current configuration data structure cache.
         String cacheName = DS_CACHE_NAME_PREFIX + cfg.getAtomicityMode() + "_" + cfg.getCacheMode() + "_" +
             cfg.getBackups() + "@" + grpName;
 
         IgniteInternalCache cache = ctx.cache().cache(cacheName);
 
+        // If it was not found or collection header does not exists in shared cache we can create separate cache.
+        if (!shared && (cache == null || cache.get(new GridCacheSetHeaderKey(dsName)) == null)) {
```

Comment:
I really don't like this "compatibility" change
```
cache.get(new GridCacheSetHeaderKey(dsName)) == null)
```
but without such search we can have "orphaned" instances of old non-collocated version.